content/blog/2025-04-24-super-optimization/index.md

```diff
@@ -0,0 +1,104 @@
+
+However, there are still situations when pouring a lot of resources into small pieces of code can be very useful, and techniques from superoptimization can influence / inspire normal compilers as well. This [overview of recent e-graph applications](https://github.com/philzook58/awesome-egraphs?tab=readme-ov-file#program-optimization) shows a bunch of recent (mainly academic) projects using e-graphs for program optimization, in varied domains, such as compiling tensor graphs, SQL queries, and GPU kernels.
+
+Not all of these are superoptimization: some just aim to apply similar techniques, but with the same goals as normal program optimization. In addition, algorithms originally developed for superoptimization have [been adapted](https://cfallin.org/pubs/egraphs2023_aegraphs_slides.pdf) for parts of the Cranelift compiler backend. On this note, it’s interesting to consider how superoptimization can be of use for normal compilers, more broadly. The paper “Automatic Generation of Peephole Superoptimizers” explored ways in which superoptimized general-purpose code snippets can be used to generate a database of transformations, and showed that it can enable signficant speed-ups for compiled code.
+
+## The inputs to Denali
+
+Denali’s approach takes these ingredients:
+
+1. A program you want to be faster
+2. A complete set of mathematical / architectural axioms (to find other equivalent programs)
+3. A cycle-accurate CPU cost model
+
+We are in no short supply of the first.
+
+For (2), an incomplete set of axioms leads only to potentially missed optimizations, and they’re a pretty one-time cost: once you have a good set of axioms for an ISA, they don’t need much maintenance. It does seem that in most cases a detailed understanding of the ISA is necessary to develop complete/correct architectural axioms. In a world where Denali-style superoptimization was more prevalent we could perhaps imagine axioms being packaged as part of the specification: there are incentives for manufacturers to enable as much optimization of code on their platforms as possible, and in many cases they seem best positioned to produce these inputs. Although incomplete axioms might only lead to missed opportunities, incorrect axioms (e.g. specifying an operation as commutative when it is not) might lead to incorrect execution, so approaches to test user-specified axioms or even synthesize them in a way that is provably correct could be worth exploring. 
+
+On the other hand, (3) seems to require quite a bit of effort: not only is designing such a cost model imposingly difficult, it also seems like an uphill battle: by the time you’re done, there’ll be a new processor to re-start the process with. The practical options seem to be either abandoning precise cycle-count as a cost model or using physical measurements to evaluate it (as Souper and STOKE do, respectively).
```

Comment:
Not sure this is worth considering, but: it is interesting to think about whether these architectural axioms or a traditional compiler backend is easier to engineer. As in, this observation:

> by the time you’re done, there’ll be a new processor to re-start the process with

…also applies to instruction selectors, scheduling models, etc. Does a Denali backend represent a similar level of effort? Easier? Harder? Who knows!